File: backend/routes/pingboard.js
import express from 'express';
const router = express.Router();
import { query } from '../db.js';

// Check individual service endpoint
async function checkService(service) {
    const startTime = Date.now();

    console.log(`Checking: ${service.url}`);
    
    try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), 10000); // 10 second timeout

        const response = await fetch(service.url, { 
            signal: controller.signal,
            headers: {
                'User-Agent': 'API-Monitor/1.0' // Good practice for APIs
            }
        });

        console.log(`Response status: ${response.status}`);
        
        clearTimeout(timeoutId);
        const responseTime = Date.now() - startTime;
        
        let status = 'up';
        if (responseTime > 2000) status = 'slow'; // 2+ seconds = slow
        if (!response.ok) status = 'down';

        // Save successful check
        await query(
            `INSERT INTO service_checks 
             (service_id, status_code, response_time, status) 
             VALUES ($1, $2, $3, $4)`,
            [service.id, response.status, responseTime, status]
        );

        return { success: true, responseTime, status: response.status };

    } catch (error) {
        const responseTime = Date.now() - startTime;

        console.error(`Fetch error details:`, error);
        
        // Save failed check
        await query(
            `INSERT INTO service_checks 
             (service_id, status_code, response_time, status, error_message) 
             VALUES ($1, $2, $3, $4, $5)`,
            [service.id, null, responseTime, 'down', error.message]
        );

        return { success: false, error: error.message };
    }
}

// Return summary info on all registered services
router.get('/services', async (req, res) => {
    try {
        const services = await query(`
            SELECT ms.*, 
                COUNT(sc.id) as total_checks,
                AVG(CASE WHEN sc.status = 'up' THEN 1 ELSE 0 END) * 100 as uptime_percentage,
                AVG(sc.response_time) as avg_response_time,
                MIN(sc.checked_at) as first_check,
                MAX(sc.checked_at) as last_check
            FROM monitored_services ms
            LEFT JOIN service_checks sc ON ms.id = sc.service_id
            WHERE ms.is_active = true
            GROUP BY ms.id
            ORDER BY ms.id
        `);
        
        res.json(services.rows);
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

// Return info on individual checks for a service
router.get('/:id/checks', async (req, res) => {
    try {
        const { id } = req.params;
        const { limit = 50 } = req.query;
        
        const checks = await query(
            `SELECT * FROM service_checks 
             WHERE service_id = $1 
             ORDER BY checked_at DESC 
             LIMIT $2`,
            [id, limit]
        );
        
        res.json(checks.rows);
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

// Check all registered services
export async function runAllChecks() {
    try {
        const services = await query(
            'SELECT * FROM monitored_services WHERE is_active = true'
        );
        
        for (const service of services.rows) {
            console.log(`Checking ${service.name}...`);
            await checkService(service);
            // Small delay between checks to be respectful
            await new Promise(resolve => setTimeout(resolve, 1000));
        }
    } catch (error) {
        console.error('Error running checks:', error);
    }
}

export default router;

File: backend/app/routes/pingboard.py
import time
import asyncio

import aiohttp
from fastapi import APIRouter, Query

from ..db import get_pool

router = APIRouter(
    prefix="/api/pingboard"
)


@router.get("/services")
async def get_services() -> list[dict]:
    pool = await get_pool()
    q = """SELECT ms.*,
                COUNT(sc.id) as total_checks,
                AVG(CASE WHEN NOT sc.status = 'down' THEN 1 ELSE 0 END) * 100 as uptime_percentage,
                AVG(sc.response_time) as avg_response_time,
                MIN(sc.checked_at) as first_check,
                MAX(sc.checked_at) as last_check,
                (SELECT status FROM service_checks
                 WHERE service_id = ms.id
                 ORDER BY checked_at DESC LIMIT 1) as latest_status,
                (SELECT checked_at FROM service_checks
                 WHERE service_id = ms.id AND status = 'down'
                 ORDER BY checked_at DESC LIMIT 1) as last_outage,
                (SELECT array_agg(rt ORDER BY ca)
                 FROM (SELECT response_time as rt, checked_at as ca
                       FROM service_checks
                       WHERE service_id = ms.id AND status != 'down'
                       ORDER BY checked_at DESC LIMIT 24) sub) as recent_response_times
            FROM monitored_services ms
            LEFT JOIN service_checks sc ON ms.id = sc.service_id
            WHERE ms.is_active = true
            GROUP BY ms.id
            ORDER BY ms.id"""
    rows = await pool.fetch(q)
    return [dict(r) for r in rows]


@router.get("/{service_id}/checks")
async def get_service_checks(service_id: int, limit: int = Query(default=50)) -> list[dict]:
    pool = await get_pool()
    q = """SELECT * FROM service_checks
            WHERE service_id = $1
            ORDER BY checked_at DESC
            LIMIT $2"""
    rows = await pool.fetch(q, service_id, limit)
    return [dict(r) for r in rows]


async def check_service(service: dict) -> tuple[int, int | None, int, str, str | None]:
    start_time = time.monotonic()

    print(f"Checking: {service['url']}")

    try:
        timeout = aiohttp.ClientTimeout(total=10)
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.get(
                service["url"],
                headers={"User-Agent": "API-Monitor/1.0"},
            ) as response:
                await response.read()
                response_time = int((time.monotonic() - start_time) * 1000)

                print(f"Response status: {response.status}")

                status = "up"
                if response_time > 2000:
                    status = "slow"
                if response.status >= 400:
                    status = "down"

                return (service["id"], response.status, response_time, status, None)

    except Exception as error:
        response_time = int((time.monotonic() - start_time) * 1000)
        print(f"Check failed, fetch error details: {error}")
        return (service["id"], None, response_time, "down", str(error))


async def run_all_checks() -> None:
    pool = await get_pool()
    rows = await pool.fetch("SELECT * FROM monitored_services WHERE is_active = true")
    services = [dict(row) for row in rows]
    results = await asyncio.gather(*(check_service(s) for s in services))

    if not results:
        return

    params = []
    value_rows = []
    for i, (service_id, status_code, response_time, status, error_message) in enumerate(results):
        offset = i * 5
        value_rows.append(f"(${offset+1}, ${offset+2}, ${offset+3}, ${offset+4}, ${offset+5})")
        params.extend([service_id, status_code, response_time, status, error_message])

    query = f"""INSERT INTO service_checks
         (service_id, status_code, response_time, status, error_message)
         VALUES {', '.join(value_rows)}"""
    await pool.execute(query, *params)
